webpack/webpack-plugin-utilities/src/Testing.ts

```diff
@@ -0,0 +1,129 @@
+import { createFsFromVolume, Volume, IFs } from 'memfs';
+import path from 'path';
+import webpack from 'webpack';
+import type { StatsCompilation as WebpackStatsCompilation } from 'webpack';
+import webpackMerge from 'webpack-merge';
+
+import type { MultiStats, Stats, Configuration, Compiler, StatsError } from 'webpack';
+
+/**
+ * @public
+ * This function generates a webpack compiler with default configuration and the output filesystem mapped to
+ * a memory filesystem. This is useful for testing webpack plugins/loaders where we do not need to write to disk (which can be costly).
+ * @param entry - The entry point for the webpack compiler
+ * @param additionalConfig - Any additional configuration that should be merged with the default configuration
+ * @param memFs - The memory filesystem to use for the output filesystem. Use this option if you want to _inspect_, analyze, or read the output
+ * files generated by the webpack compiler. If you do not need to do this, you can omit this parameter and the output files.
+ *
+ * @returns - A webpack compiler with the output filesystem mapped to a memory filesystem
+ *
+ * @example
+ * ```typescript
+ * import Testing from '@rushstack/webpack-plugin-utilities';
+ *
+ * describe('MyPlugin', () => {
+    it('should run', async () => {
+      const stats = await Testing.getTestingWebpackCompiler(
+        `./src/index.ts`,
+      );
+
+      expect(stats).toBeDefined();
+    });
+ * });
+ * ```
+ *
+ * @remarks
+ * If you want to be able to read, analyze, access the files written to the memory filesystem,
+ * you can pass in a memory filesystem instance to the `memFs` parameter.
+ *
+ * @example
+ * ```typescript
+ * import Testing from '@rushstack/webpack-plugin-utilities';
+ * import { createFsFromVolume, Volume, IFs } from 'memfs';
+ * import path from 'path';
+ *
+ * describe('MyPlugin', () => {
+ *  it('should run', async () => {
+ *    const virtualFileSystem: IFs = createFsFromVolume(new Volume());
+ *    const stats = await Testing.getTestingWebpackCompiler(
+ *      `./src/index.ts`,
+ *      {},
+ *      virtualFileSystem
+ *    );
+ *
+ *    expect(stats).toBeDefined();
+ *    expect(virtualFileSystem.existsSync(path.join(__dirname, 'dist', 'index.js'))).toBe(true);
+ *  });
+ * });
+ * ```
+ */
+export async function getTestingWebpackCompiler(
+  entry: string,
+  additionalConfig: Configuration = {},
+  memFs: IFs = createFsFromVolume(new Volume())
+): Promise<(Stats | MultiStats) | undefined> {
+  const compilerOptions: Configuration = webpackMerge(_defaultWebpackConfig(entry), additionalConfig);
+  const compiler: Compiler = webpack(compilerOptions);
+
+  compiler.outputFileSystem = memFs;
+  compiler.outputFileSystem.join = path.join.bind(path);
+
+  return new Promise((resolve, reject) => {
+    compiler.run((err, stats) => {
```

Comment:
Per the v5 documentation, you're supposed to close your compiler after running it.